### ISSUE TITLE
Improve performance when fetching large result sets with the wire driver

### ISSUE THREAD

**Issue by lucianomalta** [asfernandes/node-firebird-drivers#177]
o código abaixo

import { createWireClient } from "node-firebird-driver-wire";
import { performance } from "perf_hooks"; // Necessário em versões antigas do Node, ou omitir se usar Node 16+ global.

// O driver wire não precisa de DLL, basta instanciar direto
const client = createWireClient();

async function main() {
    let attachment;
    let transaction;

    try {
        const uri = "ip/3050:sql";

        attachment = await client.connect(uri, {
            username: "SYSDBA",
            password: "pass",
        });

        console.log("Conectado com sucesso!");

        transaction = await attachment.startTransaction();

        const sql = `
      execute block
      returns (id integer, descricao varchar(120))
      as
      begin
        --tabela
        id = 1;
        for select trim(r.rdb$relation_name) from rdb$relations r where r.rdb$relation_type in(0,3, 4,5) into descricao do suspend;
        --view
        id = 2;
        for select trim(r.rdb$relation_name) from rdb$relations r where r.rdb$system_flag = 0 and r.rdb$relation_type = 1 into descricao do suspend;
        --procedures
        id = 3;
        for select trim(p.rdb$procedure_name) from rdb$procedures p where p.rdb$system_flag = 0 into descricao do suspend;
        --functions
        id = 4;
        for select trim(f.rdb$function_name) from rdb$functions f where f.rdb$system_flag = 0 into descricao do suspend;
        --triggers
        id = 6;
        for select trim(t.rdb$trigger_name) from rdb$triggers t where t.rdb$system_flag = 0 into descricao do suspend;
        --generators
        id = 7;
        for select trim(g.rdb$generator_name) from rdb$generators g where g.rdb$system_flag = 0 into descricao do suspend;
        --udf
        id = 8;
        for select trim(g.rdb$generator_name) from rdb$generators g where g.rdb$system_flag = 0 into descricao do suspend;
        --roles
        id = 9;
        for select trim(r.rdb$role_name) from rdb$roles r where r.rdb$system_flag = 0 into descricao do suspend;
        --users
        id = 10;
        for select trim(u.sec$user_name) from sec$users u into descricao do suspend;
        --indices
        id = 11;
        for select trim(i.rdb$index_name)||'.'||trim(i.rdb$relation_name) from rdb$indices i where i.rdb$system_flag = 0 into descricao do suspend;
        --exceptions
        id = 12;
        for select trim(e.rdb$exception_name) from rdb$exceptions e where e.rdb$system_flag = 0 into descricao do suspend;
      end
    `;

        console.log("Executando a query...");
        const startTime = performance.now(); // Inicia o cronômetro

        const resultSet = await attachment.executeQuery(transaction, sql);
        const resultado = await resultSet.fetchAsObject();

        await resultSet.close();
        await transaction.commit();

        const endTime = performance.now(); // Para o cronômetro
        const tempoExecucao = (endTime - startTime).toFixed(2);

        console.log(`Resultado retornou ${resultado.length} linhas.`);
        console.log(`Tempo de execução: ${tempoExecucao} ms`);
        // console.log("Dados:", resultado); // Descomente para ver o array completo

    } catch (err) {
        console.error("Erro ao conectar ou consultar:", err);

        if (transaction && transaction.isValid) {
            await transaction.rollback();
        }
    } finally {
        if (attachment) {
            await attachment.disconnect();
            console.log("Desconectado.");
        }
        // Libera os sockets do driver wire para o processo do Node encerrar corretamente
        await client.dispose();
    }
}

main();

$ node ./t-wire.mjs 
Conectado com sucesso!
Executando a query...
Resultado retornou 5074 linhas.
Tempo de execução: 74361.80 ms
Desconectado.

import Firebird from "node-firebird";
import { performance } from "perf_hooks"; // Necessário em versões antigas do Node, ou omitir se usar Node 16+ global.

const options = {
    host: "ip",
    port: 3050,
    // Troque "sql" pelo caminho real no Linux ou Windows do servidor
    database: "sql", // Exemplo Linux
    // database: "C:\\Sistemas\\banco.fdb", // Exemplo Windows
    user: "SYSDBA",
    password: "pass
    wireCompression: false,
    writeCrypt: Firebird.WIRE_CRYPT_DISABLE,
    pluginName: Firebird.AUTH_PLUGIN_LEGACY,
};

// Promisify das funções base do node-firebird para manter a estrutura async/await
const attach = (opts) => new Promise((resolve, reject) => Firebird.attach(opts, (err, db) => err ? reject(err) : resolve(db)));
const startTransaction = (db) => new Promise((resolve, reject) => db.transaction(Firebird.ISOLATION_READ_COMMITTED, (err, tx) => err ? reject(err) : resolve(tx)));
const executeQuery = (tx, sql) => new Promise((resolve, reject) => tx.query(sql, [], (err, res) => err ? reject(err) : resolve(res)));
const commitTransaction = (tx) => new Promise((resolve, reject) => tx.commit((err) => err ? reject(err) : resolve()));
const rollbackTransaction = (tx) => new Promise((resolve, reject) => tx.rollback((err) => err ? reject(err) : resolve()));

async function main() {
    let db;
    let transaction;

    try {
        db = await attach(options);
        console.log("Conectado com sucesso!");

        transaction = await startTransaction(db);

        const sql = `
      execute block
      returns (id integer, descricao varchar(120))
      as
      begin
        --tabela
        id = 1;
        for select trim(r.rdb$relation_name) from rdb$relations r where r.rdb$relation_type in(0,3, 4,5) into descricao do suspend;
        --view
        id = 2;
        for select trim(r.rdb$relation_name) from rdb$relations r where r.rdb$system_flag = 0 and r.rdb$relation_type = 1 into descricao do suspend;
        --procedures
        id = 3;
        for select trim(p.rdb$procedure_name) from rdb$procedures p where p.rdb$system_flag = 0 into descricao do suspend;
        --functions
        id = 4;
        for select trim(f.rdb$function_name) from rdb$functions f where f.rdb$system_flag = 0 into descricao do suspend;
        --triggers
        id = 6;
        for select trim(t.rdb$trigger_name) from rdb$triggers t where t.rdb$system_flag = 0 into descricao do suspend;
        --generators
        id = 7;
        for select trim(g.rdb$generator_name) from rdb$generators g where g.rdb$system_flag = 0 into descricao do suspend;
        --udf
        id = 8;
        for select trim(g.rdb$generator_name) from rdb$generators g where g.rdb$system_flag = 0 into descricao do suspend;
        --roles
        id = 9;
        for select trim(r.rdb$role_name) from rdb$roles r where r.rdb$system_flag = 0 into descricao do suspend;
        --users
        id = 10;
        for select trim(u.sec$user_name) from sec$users u into descricao do suspend;
        --indices
        id = 11;
        for select trim(i.rdb$index_name)||'.'||trim(i.rdb$relation_name) from rdb$indices i where i.rdb$system_flag = 0 into descricao do suspend;
        --exceptions
        id = 12;
        for select trim(e.rdb$exception_name) from rdb$exceptions e where e.rdb$system_flag = 0 into descricao do suspend;
      end
    `;

        console.log("Executando a query...");
        const startTime = performance.now(); // Inicia o cronômetro

        // O node-firebird já retorna as linhas resolvidas em forma de array de objetos no callback da query
        const resultado = await executeQuery(transaction, sql);

        await commitTransaction(transaction);

        const endTime = performance.now(); // Para o cronômetro
        const tempoExecucao = (endTime - startTime).toFixed(2);

        console.log(`Resultado retornou ${resultado.length} linhas.`);
        console.log(`Tempo de execução: ${tempoExecucao} ms`);
        // console.log("Dados:", resultado); // Descomente para ver o array completo

    } catch (err) {
        console.error("Erro ao conectar ou consultar:", err);

        if (transaction) {
            await rollbackTransaction(transaction).catch(() => { });
        }
    } finally {
        if (db) {
            db.detach();
            console.log("Desconectado.");
        }
    }
}

main();

$ node ./t-node-firebird.mjs 
Conectado com sucesso!
Executando a query...
Resultado retornou 5074 linhas.
Tempo de execução: 530.47 ms
Desconectado.

Sei que o projeto está em versão beta.

Envio para ajudar testar.

**Comment by asfernandes** [comment]
Please test correction in node-firebird-driver-wire@0.0.1-beta.4

**Comment by lucianomalta** [comment]
Tested e aproved, thank you.
